File: r/globalopt/tests/testthat/test-benchmarks.R
test_that("run_benchmarks returns expected shape", {
  midpoint_opt <- function(lower, upper, budget, fn, seed) {
    set.seed(seed)
    x <- (lower + upper) / 2
    as.numeric(fn(x))
  }

  tbl <- run_benchmarks(
    dimensions = c(2L),
    budgets = c(10L),
    seeds = 1:2,
    optimizers = list(midpoint = midpoint_opt),
    include_deoptim = FALSE
  )

  expect_s3_class(tbl, "data.frame")
  expect_equal(nrow(tbl), 4L)
  expect_true(all(tbl$optimizer == "midpoint"))
  expect_true(all(tbl$dimension == 2L))
  expect_true(all(tbl$budget == 10L))
  expect_true(all(tbl$runs == 2L))
})

test_that("benchmark writers create output files", {
  df <- data.frame(
    optimizer = "midpoint",
    problem = "sphere",
    dimension = 2L,
    budget = 10L,
    runs = 2L,
    median_best = 0.0,
    best_of_runs = 0.0,
    success_rate = 1.0,
    median_seconds = 0.001,
    stringsAsFactors = FALSE
  )

  td <- tempdir()
  csv_path <- file.path(td, "bench_test.csv")
  md_path <- file.path(td, "bench_test.md")

  write_benchmark_csv(df, csv_path)
  write_benchmark_markdown(df, md_path)

  expect_true(file.exists(csv_path))
  expect_true(file.exists(md_path))
  expect_match(readLines(md_path, warn = FALSE)[1], "^\\| optimizer ")
})

File: python/tests/test_benchmarks.py
from pathlib import Path

from globalopt.benchmarks import (
    BenchmarkSummary,
    benchmark_table,
    random_search_optimizer,
    run_benchmarks,
    sphere,
    write_benchmark_csv,
    write_benchmark_markdown,
)


def test_objective_sphere_origin_is_zero() -> None:
    assert sphere([0.0, 0.0, 0.0]) == 0.0


def test_run_benchmarks_with_random_search_is_deterministic_shape() -> None:
    summaries = run_benchmarks(
        dimensions=(2,),
        budgets=(25,),
        seeds=(1, 2),
        optimizers={"random": random_search_optimizer},
        include_scipy_de=False,
    )

    # default_problems(2) yields 6 benchmark problems.
    assert len(summaries) == 6
    assert all(s.optimizer == "random" for s in summaries)
    assert all(s.dimension == 2 for s in summaries)
    assert all(s.budget == 25 for s in summaries)
    assert all(s.runs == 2 for s in summaries)


def test_write_benchmark_reports(tmp_path: Path) -> None:
    summaries = [
        BenchmarkSummary(
            optimizer="random",
            problem="sphere",
            dimension=2,
            budget=10,
            runs=2,
            median_best=0.5,
            best_of_runs=0.1,
            success_rate=0.0,
            median_seconds=0.001,
        )
    ]

    csv_path = tmp_path / "out" / "bench.csv"
    md_path = tmp_path / "out" / "bench.md"

    write_benchmark_csv(csv_path, summaries)
    write_benchmark_markdown(md_path, summaries)

    rows = benchmark_table(summaries)
    assert rows[0]["problem"] == "sphere"
    assert csv_path.exists()
    assert md_path.exists()
    assert "optimizer,problem" in csv_path.read_text(encoding="utf-8")
    assert "| optimizer | problem |" in md_path.read_text(encoding="utf-8")
